### ISSUE TITLE
Fast ipset member update leads to denied connections

### ISSUE THREAD

**Issue by rgarcia89** [projectcalico/calico#8648]
## Current Behavior
I am currently logging all denied connections in my cluster. During updates of my Redis cluster (consisting of 3 replicas), connections are logged as denied because members are removed from the Calico node ipsets too quickly. It seems Calico is faster than the application (Redis in this case) which attempts to connect to the just-terminated member. Since this member is no longer part of the ipset, any connections to it is blocked.

## Possible Solution
Delay the removal of the recently terminated pod's IP address and prevent its immediate reuse by a new pod. This ensures that connections are still handled by the existing network policies.

## Steps to Reproduce (for bugs)

1. Deploy a calico deny all policy with log before deny
2. Deploy a redis cluster (3 replica)
3. Run a rollout restart of the cluster
4. Check the logs - you will see denied connections

## Your Environment
* Calico version: v3.24.6
* Orchestrator version (e.g. kubernetes, mesos, rkt): AKS
* Operating System and version: azure linux

**Comment by fasaxc** [comment]
Thanks @rgarcia89 yes, I think we narrowed this down to:

* Having egress policy on the redis pod that says "allow to other redis pods"
* Redis using a headless service for peer detection

When we removed the egress policy, the problem didn't repro.  That made me think it was a race between Calico cleaning up the entry in the IP set and Redis learning that the peer was now gone.  So, redis kept probing the stopped peer for a few seconds.

I'm not sure we 100% verified that Calico was only removing the IP once the pod was actually terminated.  That is the design: we only consider a pod "gone" once the CNI plugin teardown has run, which is after the app in the pod has already stopped.

Ideally, apps would adapt to k8s to avoid sending traffic in this case.  Perhaps having a 30s graceful termination window but removing the peer from the cluster after 20s.  Then no new traffic would race with the removal from the IP set.

On Calico's side, we could:
* Add a grace period of a few seconds after pod teardown before we remove from IP sets.
* Add a no-reuse timer on each IP address that is longer needed so that we won't re-use that IP for another pod within the grace period.

I've wondered if the second bullet might be a good addition anyway but the first needs some thought.  Would it actually help!?  The peer pod is still gone so the packets would get dropped on the other host anyway?

**Comment by rgarcia89** [comment]
> Would it actually help!? The peer pod is still gone so the packets would get dropped on the other host anyway?
That's a good point. I believe this depends on how long existing cluster members attempt to probe the just-terminated one. In my tests (including other applications), this never exceeded 5 seconds. However, I guess making this value configurable wouldn't be a bad idea.

**Comment by fasaxc** [comment]
Can you add a 10s sleep to your post-shutdown hook for those pods?  So the daemon has stopped but the pod will stay up another 10s before we clean up IPs.

**Comment by rgarcia89** [comment]
I wouldn't know how to do that. There is only a preStop command that can be passed. This however wouldn't be the correct moment when focusing on the lifecycle of the pod.

**Comment by fasaxc** [comment]
Ah, there's only `preStop`.  Hmm; I do think anything we could do on the dataplane side would be a bit of a hack.  The best option I can think of is to add a delay to the CNI DEL call so that the networking for the pod stays up for a while after the process inside the pod has gone but that's really a workaround.  The best option would be for Redis to shut down in such a way that it stops sending traffic to the terminating pod _before_ the pod actually terminates.

**Comment by rgarcia89** [comment]
I completely understand your point. However, the issue isn't limited to Redis; it also occurs with Vault and our PostgreSQL cluster deployments.

**Comment by rgarcia89** [comment]
@fasaxc is there any update or idea in regards to this topic?

**Comment by fasaxc** [comment]
Only easy option I can think of is to provide a way to delay the CNI DEL operation from finishing.  That'd extend the time between binary inside the pod being stopped and networking being torn down.  Perhaps an annotation on the pod to trigger a certain delay.

**Comment by rgarcia89** [comment]
On the application pod itself you mean? I would prefer a more centralized approach to be honest. A parameter that could be placed in some calico pod configmap for instance.

**Comment by fasaxc** [comment]
Can try it, here's a test image (based on v3.27) with a 20s sleep if you want to give it a go: `fasaxc/calico-cni:v3.27-with-delay`  Just replace the `calico/cni` image with that (on a test cluster! I haven't tested it myself).

**Comment by rgarcia89** [comment]
@fasaxc I do not see any delay happening. The pod IP gets deleted in the same moment from the ipset as the pod gets deleted from the cluster. 

I have attached a recording of the test. Same happens if I terminate the pod using a grace period of 0.
[Screencast from 2024-04-23 14-36-47.webm](https://github.com/projectcalico/calico/assets/4134730/454a8579-a84e-4826-8ec4-fbe5331f8d22)

**Comment by fasaxc** [comment]
> I do not see any delay happening. The pod IP gets deleted in the same moment from the ipset as the pod gets deleted from the cluster.

Yes, that's still the case, but the pod IP should only be deleted 20s after the redis process is killed.  Perhaps that's not enough if redis still thinks that the other cluster member exists!

**Comment by rgarcia89** [comment]
I would also like to mention that this issue is not only occurring with Redis. We are also experiencing it with other applications. For example, Grafana has caused some denied logs when the pods were terminated but the user session was sticky. We observed deny logs indicating that Traefik was unable to route traffic to Grafana, which is located in the same cluster. Same issue has also been observed for Vault, Postgresql and other applications as well.

**Comment by fasaxc** [comment]
It's frustrating and hard to fix from Calico's side.  On slack we investigated the tear down sequence but the problem is that Calico can only block the CNI teardown stage, whereas Redis et al are using the headless service to identify their peers and the headless service only gets updated to remove the IP *after* the CNI teardown.  It doesn't matter how much we delay, it'll delay the cluster from updating its peers for same amount.  Really need a Kubernetes enhancement to the `publishNotReadyAddresses` to make it stop publishing once the pod is terminated (before CNI teardown is done).

**Comment by fasaxc** [comment]
I don't really see a good way to progress this one so I'm going to close for now.  I just don't think we have the time/clout to get that type of change upstream.  Ping me if you want be to reopen.

**Comment by fasaxc** [comment]
Perhaps one to raise with https://github.com/kubernetes/community/tree/master/wg-node-lifecycle
